### PR TITLE
Support opening file paths given relative to home directory with tildes

### LINKED ISSUE #12778: Can't open paths beginning with ~
### Plan

https://github.com/astropy/astropy/pull/13107#issuecomment-1104552457

### Description
`astropy.io.fits` can't open files when the path is given with a `~` representing the home directory, e.g. `fits.getdata("~/data/file.fits")`. Instead, the home directory must be spelled out, e.g. `fits.getdata("/home/sam/data/file.fits")`. This can be fixed by passing paths to `os.path.expanduser`, which should do the right thing across operating systems.

There was some discussion about `expanduser` in #10705, and it looks like the function is now in use for finding the Astropy configuration directory.

There is also [precedent in Sunpy](https://github.com/sunpy/sunpy/blob/main/sunpy/util/io.py#L26), where `expanduser` (as wrapped by `pathlib`) is used for paths passed to Sunpy's `Map` object, so that `sunpy.map.Map("~/data/file.fits")` works. (`Map` is their general object for loading, manipulating and plotting data).

If there's not a reason to avoid adding calls to `os.path.expanduser`, I'm happy to dig into the code and prepare a PR. I'm most interested in the FITS loader, but I might take a look at the rest of `astropy.io`. (The Github search tool doesn't show any uses of `expanduser` there.)

### Expected behavior
`astropy.io.fits.getdata("~/data/file.fits")` loads the file `file.fits` in the `data` directory within the user's home directory.

### Actual behavior
`astropy.io.fits.getdata("~/data/file.fits")` raises `FileNotFoundError: [Errno 2] No such file or directory: '~/data/file.fits'`

### Steps to Reproduce

Attempt to load a FITs file in the user's home directory, giving a path beginning with a `~` representing the path to the user's home directory.

### System Details
Linux-5.13.0-27-generic-x86_64-with-glibc2.31
Python 3.10.1 | packaged by conda-forge | (main, Dec 22 2021, 01:39:05) [GCC 9.4.0]
Numpy 1.22.1
pyerfa 2.0.0.1
astropy 5.1.dev516+g20513ae79
Scipy 1.7.3
Matplotlib 3.5.1

### PR DESCRIPTION
**Pull request by svank**
### Description

This PR is hopefully everything needed to close #12778 and allow paths to be given relative to the user's home directory as `~/dir/filename.fits`. I think this covers every part of the `io` module that should be covered (based on my scrolling through the docs---I could have missed something).

In most submodules within `io`, there's just some sort of read and write method, so it was pretty easy to add a call to `os.path.expanduser` to each and add a simple test. `io.fits` is by far the largest affected API. I tried adding the `expanduser` call at the bottom-most level in `fileobj_open` as discussed in #12778, but I found that there were other points in the call stack where the file path does get used (e.g. the `overwrite` flags supported by some of the output methods), and so instead I've added `expanduser` calls early in the call stack, so that the tilde is expanded before anything tries to use the file path.

At a few points in `io.fits`, functions can accept a file path or an open file object, and so the `os.path.expanduser` call comes after some sort of check to see if the input is a type that would represent a path. I noticed those checks weren't all consistent. Some use the `path_like = (str, os.PathLike)` tuple defined in `fits/utils.py`, others check for just a `str`, and a few check for a `str` or `bytes`. I didn't see a reason for this varied checking, so I added `bytes` to `path_like` and, wherever I was already making a change, I changed things to use `path_like`. In some places my change came after an existing check which I modified, but in others I had to add a new check and would have had to make a decision anyway on how to do this check. I won't claim to have harmonized this type checking over all of `io.fits`, and I can back out or separate this if it's too unrelated.

For testing tilde paths, I used pytest's `monkeypatch` to temporarily update the environment variables so that `~` resolves to either the test data directory or the temporary directory, as appropriate, and only for specific test cases. Then, depending on what I found in the existing tests, I either added a test case just for exercising `~` expansion, or I augmented an existing test to run both with data/temp paths given normally, and with data/temp paths given via `~` paths. In `io.fits` there's some existing infrastructure with a `FitsTestCase` class with methods to access the data and temp directories, and I've plugged into this to make it easy to sprinkle these tilde-path tests throughout the existing tests. I scrolled through the `io.fits` documentation to find everything that looked IO-related and end-user-facing. When I found a relatively simple test exercising those functions I added a fixture to run that test both with and without tilde paths, and otherwise I added a new test case specifically to exercise that function. There were a few utility functions that I don't think were being exercised at all in testing, so this PR might increase coverage a tad, though these new tests are more just about ensure a FileNotFound error isn't raised with tilde paths.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by mwcraig — COMMENTED** [review]
Thanks for this @svank -- it takes considerable effort to dive into `io.firts` internals. I'm only reviewing the `nddata` bit here: that looks good.

File: astropy/utils/data.py [hunk 1]
```diff
@@ -268,6 +268,7 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
                 name_or_obj, cache=cache, show_progress=show_progress,
                 timeout=remote_timeout, sources=sources,
                 http_headers=http_headers)
+        name_or_obj = os.path.expanduser(name_or_obj)
```
Comment (pllim) [inline astropy/utils/data.py:271]: This is unnecessary if the file is downloaded from URL?
```suggestion
        else:
            name_or_obj = os.path.expanduser(name_or_obj)
```

**Review by pllim — COMMENTED** [review]
I wonder if this PR should be split up into one per subpackage.

Regardless, it needs one change log per subpackage, where applicable.

**Comment by svank** [comment]
I'd be happy to split this up if that's the consensus. The commits break down per-subpackage pretty well so it shouldn't be hard, and I did feel like I was throwing out a lot all at once after I sent the PR.

I'll add changelogs once that's decided.

File: astropy/utils/data.py (hunk 1 above)
Comment (svank) [reply astropy/utils/data.py]: Yes, good point

**Comment by pllim** [comment]
@svank , if you don't mind, it is easier to merge some of the non-controversial ones if you split it up. FITS is never non-controversial...

**Comment by svank** [comment]
@pllim Looks like the one change in `astropy/utils/data.py` (the change you commented on) is included in the `io.ascii` commit but is also a dependency of the `io.votable` changes and the `nddata` tests. Meanwhile `nddata` also depends on the `io.fits` changes, and the `table` commit is stand-alone. To work through the dependency chain, I'll start with PRs for `io.ascii`, `io.fits`, and `table`, and then submit `io.votable` and `nddata` as things get merged.

**Comment by svank** [comment]
@pllim Looks like the one change in `astropy/utils/data.py` (the change you commented on) is included in the `io.ascii` commit but is also a dependency of the `io.votable` changes and the `nddata` tests. Meanwhile `nddata` also depends on the `io.fits` changes, and the `table` commit is stand-alone. To work through the dependency chain, I'll start with PRs for `io.ascii`, `io.fits`, and `table`, and then submit `io.votable` and `nddata` as things get merged.

**Comment by pllim** [comment]
Sounds like a plan. Sorry for the extra work and thanks for your patience!

**Comment by pllim** [comment]
This PR has been split into smaller PRs, so I am closing this one. Thanks again!

**Pull request closed without merging**